python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py

```diff
@@ -0,0 +1,156 @@
+        self.__batches = None
+        self.__sample_call_args = kwargs
+
+    @property
+    def rank(self) -> int:
+        return self.__rank
+
+    @property
+    def saturation_level(self) -> int:
+        return self.__saturation_level
+
+    @property
+    def size(self) -> int:
+        if self.__batches is None:
+            return 0
+        else:
+            return len(self.__batches)
+
+    def add_batches(
+        self,
+        df: Union[cudf.DataFrame, dask_cudf.DataFrame],
+        start_col_name: str,
+        batch_col_name: str,
+    ) -> None:
+        """
+        Adds batches to this BulkSampler.
+
+        Parameters
+        ----------
+        df: cudf.DataFrame or dask_cudf.DataFrame
+            Contains columns for vertex ids, batch id
+        start_col_name: str
+            Name of the column containing the start vertices
+        batch_col_name: str
+            Name of the column containing the batch ids
+
+        Returns
+        -------
+        """
+        df = df.rename(
+            columns={
+                start_col_name: self.start_col_name,
+                batch_col_name: self.batch_col_name,
+            }
+        )
+
+        if self.__batches is None:
+            self.__batches = df
+        else:
+            if isinstance(df, type(self.__batches)):
+                self.__batches = self.__batches.append(df)
+            else:
+                raise TypeError(
+                    "Provided batches must match the dataframe"
+                    " type of previous batches!"
+                )
+
+        if self.size >= self.saturation_level:
+            self.flush()
+
+    def flush(self) -> None:
+        """
+        Computes all uncomputed batches
+        """
+        end = min(self.__saturation_level, len(self.__batches))
+
+        sample_fn = (
+            cugraph.dask.uniform_neighbor_sample
+            if isinstance(self.__batches, dask_cudf.DataFrame)
+            else cugraph.uniform_neighbor_sample
+        )
+
+        # TODO semaphore check to prevent concurrent calls to uniform_neighbor_sample
+        samples = sample_fn(
+            self.__graph,
+            **self.__sample_call_args,
+            start_list=self.__batches[self.start_col_name][:end],
+            batch_id_list=self.__batches[self.batch_col_name][:end],
+            with_edge_properties=True,
+        )
+
+        if len(self.__batches) > end:
+            self.__batches = self.__batches[end:]
+        else:
+            self.__batches = None
+
+        self.__write(samples)
+
+    def __write(self, samples: Union[cudf.DataFrame, dask_cudf.DataFrame]) -> None:
+        # Ensure each rank writes to its own partition so there is no conflict
+        outer_partition = f"rank={self.__rank}"
+        if isinstance(samples, dask_cudf.DataFrame):
+            samples.to_parquet(
+                os.path.join(self.__output_path, outer_partition),
+                partition_on=["batch_id", "hop_id"],
+            )
```

Comment:
I suggest only partitioning on a single `batch_id` , I am afraid that we might end up with too small files which can  cause  bottlenecks on the IO layer.  

We might want to benchmark what is the fastest way to read each batch. I think we might have to have batches that are at least a few gigs to make this IO not be slow.  

We might have
```suggestion
            samples.to_parquet(
                os.path.join(self.__output_path, outer_partition),
                partition_on=["batch_id"],
            )
```

Reply:
ok